lib/spack/spack/environment.py

```diff
@@ -337,47 +395,62 @@ def __init__(self, path, init_file=None):
             path (str): path to the root directory of this environment
             init_file (str or file object): filename or file object to
                 initialize the environment
+            with_view (str or bool): whether a view should be maintained for
+                the environment. If the value is a string, it specifies the
+                path to the view.
         """
         self.path = os.path.abspath(path)
         self.clear()
 
         if init_file:
-            # initialize the environment from a file if provided
             with fs.open_if_filename(init_file) as f:
                 if hasattr(f, 'name') and f.name.endswith('.lock'):
-                    # Initialize the environment from a lockfile
+                    self._read_manifest(default_manifest_yaml, init=True)
                     self._read_lockfile(f)
                     self._set_user_specs_from_lockfile()
-                    self.yaml = _read_yaml(default_manifest_yaml)
                 else:
-                    # Initialize the environment from a spack.yaml file
-                    self._read_manifest(f)
+                    self._read_manifest(f, init=True)
         else:
-            # read lockfile, if it exists
+            init = not any(os.path.exists(x)
+                           for x in (self.lock_path, self.manifest_path))
+            default_manifest = not os.path.exists(self.manifest_path)
+            if default_manifest:
+                self._read_manifest(default_manifest_yaml, init=init)
+            else:
+                with open(self.manifest_path) as f:
+                    self._read_manifest(f, init=init)
+
             if os.path.exists(self.lock_path):
                 with open(self.lock_path) as f:
                     self._read_lockfile(f)
+                if default_manifest:
+                    self._set_user_specs_from_lockfile()
 
-            if os.path.exists(self.manifest_path):
-                # read the spack.yaml file, if exists
-                with open(self.manifest_path) as f:
-                    self._read_manifest(f)
-
-            elif self.concretized_user_specs:
-                # if not, take user specs from the lockfile
-                self._set_user_specs_from_lockfile()
-                self.yaml = _read_yaml(default_manifest_yaml)
-            else:
-                # if there's no manifest or lockfile, use the default
-                self._read_manifest(default_manifest_yaml)
+        if with_view is False:
+            self._view_path = None
+        elif isinstance(with_view, six.string_types):
+            self._view_path = with_view
+        # If with_view is None, then defer to the view settings determined by
+        # the manifest file
 
-    def _read_manifest(self, f):
+    def _read_manifest(self, f, init=False):
```

Comment:
I think this parameter makes the calls to this function hard to understand.  What's `init` supposed to mean here?

Reply:
@tgamblin `init` determines whether the environment already exists (on the filesystem) or is being created here: in the latter case. The intended behavior is the following

* for newly-created environments, the default is to create a view
* for existing environments the default is explicitly disable a view unless the user explicitly requests it (for example if they set `view: True` in the `spack.yaml`)

You were of the opinion that views should always be enabled right? This seems workable but I will need to think through the possible repercussions more.